File: JavaScript/08_Greedy Algorithm/M-GFG-Job Sequencing Problem.js
// GFG: Job Sequencing Problem
// https://www.geeksforgeeks.org/problems/job-sequencing-problem-1587115620/1
// Company Tags: Flipkart, Accolite, Microsoft

/**
 * @param {number[]} deadline
 * @param {number[]} profit
 * @return {number[]}
 */
function jobSequencing(deadline, profit) {
    const n = deadline.length;

    // Store {profit, deadline, jobId}
    const jobs = [];
    for (let i = 0; i < n; i++) {
        jobs.push({ profit: profit[i], deadline: deadline[i], id: i });
    }

    // Sort jobs by profit descending
    jobs.sort((a, b) => b.profit - a.profit);

    // Find max deadline
    let maxDeadline = 0;
    for (const d of deadline) {
        maxDeadline = Math.max(maxDeadline, d);
    }

    // DSU parent array
    const parent = [];
    for (let i = 0; i <= maxDeadline; i++) {
        parent[i] = i;
    }

    // Find parent with path compression
    function find(s) {
        if (parent[s] === s) return s;
        return parent[s] = find(parent[s]);
    }

    let totalProfit = 0;
    let totalJobs = 0;

    for (const job of jobs) {
        const d = job.deadline;
        const p = job.profit;

        // Find available slot <= d
        const availableSlot = find(d);
        if (availableSlot > 0) {
            totalProfit += p;
            totalJobs++;
            // Union: mark this slot as filled -> point to previous slot
            parent[availableSlot] = find(availableSlot - 1);
        }
    }

    return [totalJobs, totalProfit];
}

const deadline = [4, 1, 1, 1];
const profit = [20, 10, 40, 30];
console.log(jobSequencing(deadline, profit)); // Output: [2, 60]

File: Java/08_Greedy Algorithm/M-GFG-Job Sequencing Problem.java
// GFG: Job Sequencing Problem
// https://www.geeksforgeeks.org/problems/job-sequencing-problem-1587115620/1
// Company Tags: Flipkart, Accolite, Microsoft

import java.util.*;

class Solution {

    // Find parent (with path compression)
    private int find(int parent[], int s) {
        if (parent[s] == s) return s;
        return parent[s] = find(parent, parent[s]);
    }

    public ArrayList<Integer> jobSequencing(int[] deadline, int[] profit) {
        int n = deadline.length;

        // store {profit, deadline, jobId}
        int[][] jobs = new int[n][3];
        for (int i = 0; i < n; i++) {
            jobs[i][0] = profit[i];
            jobs[i][1] = deadline[i];
            jobs[i][2] = i;
        }

        // Sort jobs by profit descending
        Arrays.sort(jobs, (a, b) -> b[0] - a[0]);

        // Find max deadline
        int maxDeadline = 0;
        for (int d : deadline) {
            maxDeadline = Math.max(maxDeadline, d);
        }

        // DSU parent array
        int parent[] = new int[maxDeadline + 1];
        for (int i = 0; i <= maxDeadline; i++) parent[i] = i;

        int totalProfit = 0,
            totalJobs = 0;

        for (int[] job : jobs) {
            int d = job[1];
            int p = job[0];

            // Find available slot <= d
            int availableSlot = find(parent, d);
            if (availableSlot > 0) {
                totalProfit += p;
                totalJobs++;
                // Union: mark this slot as filled -> point to previous slot
                parent[availableSlot] = find(parent, availableSlot - 1);
            }
        }

        ArrayList<Integer> result = new ArrayList<>();
        result.add(totalJobs);
        result.add(totalProfit);
        return result;
    }
}
